presto-hudi/src/main/java/com/facebook/presto/hudi/HudiFileSkippingManager.java

```diff
@@ -0,0 +1,312 @@
+import static com.facebook.presto.parquet.predicate.PredicateUtils.isStatisticsOverflow;
+import static java.lang.Float.floatToRawIntBits;
+
+/**
+ * MetaDataTable based filesManager.
+ * list Hudi Table contents based on the
+ *
+ * <ul>
+ *   <li>Table type (MOR, COW)</li>
+ *   <li>Query type (snapshot, read_optimized, incremental)</li>
+ *   <li>Query instant/range</li>
+ * </ul>
+ *
+ * Support dataSkipping.
+ */
+public class HudiFileSkippingManager
+{
+    private static final Logger log = Logger.get(HudiFileSkippingManager.class);
+
+    private final HoodieTableQueryType queryType;
+    private final Option<String> specifiedQueryInstant;
+    private final HoodieTableMetaClient metaClient;
+    private final HoodieTableType tableType;
+    private final HoodieTableMetadata metadataTable;
+
+    protected transient volatile Map<String, List<FileSlice>> allInputFileSlices;
+
+    private transient volatile SyncableFileSystemView fileSystemView;
+
+    public HudiFileSkippingManager(
+            List<String> partitions,
+            String spillableDir,
+            HoodieEngineContext engineContext,
+            HoodieTableMetaClient metaClient,
+            HoodieTableQueryType queryType,
+            Option<String> specifiedQueryInstant)
+    {
+        this.queryType = queryType;
+        this.specifiedQueryInstant = specifiedQueryInstant;
+        this.metaClient = metaClient;
+        HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder().enable(true).build();
+        this.tableType = metaClient.getTableConfig().getTableType();
+        this.metadataTable = HoodieTableMetadata
+                .create(engineContext, metadataConfig, metaClient.getBasePathV2().toString(), spillableDir, true);
+        prepareAllInputFileSlices(partitions, engineContext, metadataConfig, spillableDir);
+    }
+
+    private void prepareAllInputFileSlices(List<String> partitions, HoodieEngineContext engineContext, HoodieMetadataConfig metadataConfig, String spillableDir)
+    {
+        long startTime = System.currentTimeMillis();
+        HoodieTimeline activeTimeline = metaClient.reloadActiveTimeline();
+        Option<HoodieInstant> latestInstant = activeTimeline.lastInstant();
+        // build system view.
+        fileSystemView = FileSystemViewManager
+                .createViewManager(engineContext, metadataConfig,
+                        FileSystemViewStorageConfig.newBuilder().withBaseStoreDir(spillableDir).build(),
+                        HoodieCommonConfig.newBuilder().build(),
+                        () -> metadataTable)
+                .getFileSystemView(metaClient);
+        Option<String> queryInstant = specifiedQueryInstant.or(() -> latestInstant.map(HoodieInstant::getTimestamp));
+        if (tableType.equals(HoodieTableType.MERGE_ON_READ) && queryType.equals(HoodieTableQueryType.SNAPSHOT)) {
+            allInputFileSlices = partitions.stream().collect(Collectors.toMap(Function.identity(),
```

Comment:
can we leverage engineContext and parallelize the calls? i.e. we could fetch file slices concurrently for diff partitions instead of doing it sequentially.

Reply:
thanks
i also find this function take too long.  1.5w files takes 700ms